README
================================
SWOT Simulator for Ocean Science
================================

Description
-----------
This software simulates SWOT sea surface height (SSH) observations that can be applied to an ocean general circulation model (OGCM), allowing the exploration of ideas and methods to optimize information retrieval from the SWOT Mission in the future. From OGCM SSH inputs, the software samples SWOT-like outputs on a swath along the orbit ground track and adds measurement error and noise, which are generated according to technical characteristics published by the SWOT project team.
The simulator can also be used to simulate nadir-like data for any altimeters.

The swotsimulator code can be downloaded at http://swot.jpl.nasa.gov/science/resources/

Licence
-------
The swotsimulator is under BSD licence (see :ref:`COPYING <licence_long>` )

Installation
------------
The code runs with python and uses the following libraries:

* Numpy

* Scipy

* NetCDF4 if you need to read netcdf4 model files (not included in Canopy)

If you don't have python and the needed python libraries on your machine
you can download the enthought python distribution canopy at
<https://store.enthought.com/> and follow the installation instructions.

To install swotsimulator:

-> If you have installed canopy:

Open a canopy terminal (``Tools - Canopy Terminal``)

.. code-block:: python

   > cd [yourpath]/swotsimulator/
   > python setup.py install

->If you have your own python and libraries:

- global installation (requires root privilege):

.. code-block:: python

   > sudo python setup.py install

- local installation:

.. code-block:: python

   > python setup.py install --home=your/local/path

- Uninstall:

.. code-block:: python

   > sudo python setup.py install --record record_files.txt
   > cat record_files.txt | xargs sudo rm -rf

Running
--------
Run the SWOT simulator:

.. code-block:: python

   > swotsimulator [your params file]

Run the nadir alone:

.. code-block:: python

   > nadirsimulator [your params file]


For example, to run the example do:

* open ``./example/params_example.py`` and check that the directory path
  are correct, if needed modify path of ``dir_setup``, ``indatadir``, ``outdatadir``
  with your own path (the path of the simulator directory).

* Run the simulator:
  ``swotsimulator ./example/params_example.py``


Note that if you have installed python using enthought Canopy, you should either
use the Canopy terminal or run the canopy python instead of python (direct path or
use an alias).

The data provided in the example were produced by the Regional Ocean Modeling
 System (ROMS) off the Oregon coast developed by Dr. Yi Chao and his team.

Testing
-------
Outputs of the simulator can be tested using algorithm from the test directory. 

To run all tests, go to the test directory:

.. code-block:: python

  > python run_all.py [your params file]

Documentation
-------------
- To build the documentation, in the ``doc`` directory:

    * Build html: ``make html``

    * Build pdf: ``make latexpdf``

  The build documentation files are located in ``doc/build/html`` and in ``doc/build/latex/``

- for a complete description:
  see the ``doc`` directory or just run ``pydoc PyDom``

  .. doc: doc/html/index.html



CHANGES.txt
Changelog for swotsimulator
===========================



This file lists the changes in each swotsimulator version
3.2
---
06/29/2018
- Implement three outputs format: classic for the old one, mockup to generate
SWOT-like future L2 data, expert to save only noise in 1D.


3.1
---
04/26/2018
- Implement Multiprocessing for the swotsimulator
- User-proof Random coefficient and savesignal options
- Update documentation (detail how to use coeff_file and savesignal options
- Fix bug with numpy 14.2
- Fix bug with interpolation at 0E
and add multiprocessing option)

3.0
---
12/27/2017
- Improve packaging, swotsimulator have to be installed
- Fix issues on mask when interpolating using pyresample
- Add variables and change type according to inputs from CNES/JPL format sheet
- Add expert mode, to save only noise in 1D

2.3
-----
08/31/2017
- Decorrelate left and right part of the mast for timing error
- Add option to generate random signals instead of random coefficients
- Handle mask when netCDF file does not have one
- Resolve interpolation issues for Global model around 0degE

07/31/2017
- Add option number of columns to read any orbit files

06/26/2017
- Add digit precision in time variable (from f4 to f8)

03/15/2017
- Fix bug in interpolation
03/01/2017
- Fix bug when running the simulator without nadir altimeter
- Improve simulator with PEP8 convention and logger debug message
- User proof and clean code

02/22/2017
- Fix bug running the simulator to simulate along track altimeters alone

12/19/2016
- Updating Orbit File with those from http://www.aviso.altimetry.fr/en/missions/future-missions/swot/orbit.html. Two new files have been gitted: ephem_science_sept2015_ell.txt for the science orbit and ephem_calval_june2015_ell. for the 1-day (for calval purpose) orbit.

2.2.1
-----
11/20/2016
- Add plotting script in the documentation to automatically update figure with outputs from the last version of the simulator.

08/20/2016
- Bug fixed:
        -> Spatial lag (params key ... ) was not implemented properly

2.2
-----
09/25/2016
- Upgrade of the code to be pep8 compliant.

09/23/2016
- Adding new variables and global attribute to improve CF1.6 compliancy of netcdf outputs from the simulator


08/20/2016
- Minor modification:
        -> Karin noise is updated with the version provided by th SWOT technical team, swath is now from 5km to 60 from Nadir and SWH can go up to 8~m.

07/29/2016
- Add warnings into rw_data to user-proof the code.

2.1
-----
06/15/2016
Version available on github

2.0.0
------
04/15/2016

- Major modifications:

        -> The code is now compatible with Python3

        -> There are now 2 numbers to define random realisations in the parameter file: ncomp1d for the number of 1d realisations (all the instrument errors) and ncomp2d for the number of 2d realisations (wet-tropospheric error). This enable the user to increase the number of 1d realisations for longer wavelengths without increasing the computing time for the computation of the wet-tropospheric error.   
 
        -> Nadir data are saved in a separate file for the grid and for the observation-like outputs

        -> Passes are now splitted in ascending (odd numbers) and descending (even number) passes.

        -> Netcdf format of outputs is improved

        -> Interpolations of the model on SWOT swath have evolved so that the simulator can run faster: If the model grid is regular (options grid in parameter file), spline interpolation can be used. Regular or irregular grid can be interpolated using pyresample if it is installed or griddata (from scipy.interpolate). Note that griddata can be very slow or even run out of memory for large number of data. 

- Minor modifications:  

        -> Main scripts run_simulator.py and run_nadir.py have been rewritten to make the code more flexible and easily readable.

        -> Improve printing of error messages and detection of potential errors

        -> The nadir instrument errors were a random noise. It has been modified so that the nadir instrument errors are now following the spectrum of errors of current nadir altimeters. 

        -> Orbit interpolated at high resolution if the provided orbit has a low temporal resolution (more than 0.5s)

1.0.0
------
10/31/2014

- Major modifications:

	-> Implementation of SWH (Sea Wave Height) dependant KaRIN noise: New file karin_noise.nc added to data. The value of SWH is specified in the params file and is constant in time and space. 

	-> Save normalized KaRIN noise every nrandkarin km, and random coefficient for other noises in file_coeff specified in params so that runs are reproducible (random coefficients are loaded from that file when it exists). 

	-> Set up an option to compute the noise alone if no model file is provided (set 'file_input' to None in params)

	-> Karin noise is now define on a swath from 10km to 60 km from the nadir. The parameter halfgap should be change to 10km in the params.txt file

	-> Compute data at the nadir using a noise similar to the altimeter ... WARNING: the along track sampling is the same as the one in SWOT data. 

	-> Possibility to compute nadir-like data for any altimeter using run_nadir.py. It runs exactly like run.py. An example of the parameter file is in example/params_example_nad.txt, the along track sampling can be set in the params file (recommended value is 6km). 

- Minor modifications:

	-> Number of random frequencies is now coded as a parameter and can be modified in params. 

	-> Change name of variable indice by index in output file. 

	-> renamed the module to swotsimulator, according to Python's PEP standard naming convention

	-> Reading Netcdf4 files is now possible if netCDF4 python module is installed
	
- Fixed bugs:
	-> problem around 0deg longitude due to the discontinuity 0-360 is fixed

	-> Last pass missing: fixed

1.0 beta 1
----------
09/01/2014 



example/params_example.py
# -----------------------#
# Files and directories
# -----------------------#
## -- Get the user home directory
from os.path import expanduser
import os
home = expanduser("~")
# ------ Directory that contains orbit file:
dir_setup = os.path.join(home, 'swotsimulator', 'data')
# ------ Directory that contains your own inputs:
indatadir = os.path.join(home, 'swotsimulator', 'example',
                         'input_fields')
# ------ Directory that contains your outputs:
outdatadir = os.path.join(home, 'swotsimulator', 'example',
                           'swot_output')
# ------ Orbit file:
# Order of columns (lon, lat, time) in the orbit file
# (default is (1, 2, 0) with order_orbit_col = None)
order_orbit_col = None
# Name of the orbit file
satname = "science"
filesat = dir_setup + os.sep + 'ephem_science_sept2015_ell.txt'
# ------ Name of the configuration (to build output files names)
config = "OREGON"
#Number of processors to be used
proc_number = 3

# -----------------------#
# SWOT swath parameters
# -----------------------#
# ------ Satellite grid file root name:
# 	 (Final file name is root_name_[numberofpass].nc)
filesgrid = os.path.join(outdatadir, '{}_{}_grid'.format(config,satname))
# ------ Force the computation of the satellite grid:
makesgrid = True
# ------ Give a subdomain if only part of the model is needed:
#	 (modelbox=[lon_min, lon_max, lat_min, lat_max])
# 	 (If modelbox is None, the whole domain of the model is considered)
modelbox = None  # [230.144,234.598,42.27,47.8283]
# ------ Distance between the nadir and the end of the swath (in km):
halfswath = 60.
# ------ Distance between the nadir and the beginning of the swath (in km):
halfgap = 10.
# ------ Along track resolution (in km):
delta_al = 2.
# ------ Across track resolution (in km):
delta_ac = 2.
# ------ Shift longitude of the orbit file if no pass is in the domain
#        (in degree): Default value is None (no shift)
shift_lon = None
# ------ Shift time of the satellite pass (in day):
#        Default value is None (no shift)
shift_time = None

# -----------------------#
# Model input parameters
# -----------------------#
# ------ List of model files:
#	 (The first file contains the grid and is not considered as model data)
#        To generate the noise alone, file_input = None
#        and specify region in modelbox
file_input = os.path.join(indatadir, 'list_of_file.txt')
# ------ Type of model data:
#	 (Optional, default is NETCDF_MODEL and reads netcdf3 and netcdf4 files)
#	 (Other options are ROMS, NEMO and CLS to read Nemo, roms or CLS)
model = 'NETCDF_MODEL'
# ------ Type of grid:
#        'regular' or 'irregular', if 'regular' only 1d coordinates
#        are extracted from model
grid = 'irregular'
# ------ Specify SSH variable:
var = 'H'
# ------ Specify factor to convert SSH values in m:
SSH_factor = 1.
# ------ Specify longitude variable:
lon = 'lon_rho'
# ------ Specify latitude variable:
lat = 'lat_rho'
# ------ Time step between two model outputs (in days):
timestep = 1.
# ------ Number of outputs to consider:
#        (timestep*nstep=total number of days)
nstep = 25.
# ------ Not a number value:
model_nan = 0.

# -----------------------#
# SWOT output files
# -----------------------#
# ------ Output file root name:
#	 (Final file name is root_name_c[cycle]_p[pass].nc
file_output = os.path.join(outdatadir, '{}_{}'.format(config, satname))
# ------ Interpolation of the SSH from the model (if grid is irregular and
#         pyresample is not installed:
#        (either 'linear' or 'nearest', use 'nearest' for large region
#        as it is faster and use less memory.)
interpolation = 'linear'
# ------ Save variables with mockup variables ('mockup') swotsimulator
#        variables ('classic', default behaviour) or in expert mode ('expert')
save_variables = 'classic'

# -----------------------#
# SWOT error parameters
# -----------------------#
# ------ File containing random coefficients to compute and save
#	 random error coefficients so that runs are reproducible:
#        If file_coeff is specified and does not exist, file is created
#	 If you don't want runs to be reproducible, file_coeff is set to None
file_coeff = None
# file_coeff = os.path.join(outdatadir,'Random_coeff.nc')
# ------ KaRIN noise (True to compute it):
karin = True
# ------ KaRIN file containing spectrum for several SWH:
karin_file = os.path.join(dir_setup, 'karin_noise.nc')
# ------ SWH for the region:
#        if swh greater than 7 m, swh is set to 7
swh = 2.0
# ------ Number of km of random coefficients for KaRIN noise (recommended nrandkarin=1000):
nrandkarin = 1000

## -- Other instrument error (roll, phase, baseline dilation, timing)
## -----------------------------------------------------------------
# -- Compute nadir (True or False):
nadir = True
# ------ File containing spectrum of instrument error:
file_inst_error = os.path.join(dir_setup, "global_sim_instrument_error.nc")
# ------ Number of random realisations for instrumental and geophysical error
#        (recommended ncomp=2000), ncomp1d is used for 1D spectrum, and ncomp2d
#        is used for 2D spectrum (wet troposphere computation):
ncomp1d = 4000
ncomp2d = 2000
# ------ Cut off frequency:
#	 (Use lambda_cut=40000km for cross-calibration)
lambda_cut = 20000
lambda_max = lambda_cut
# ------ Save entire rando signal instead of random coefficients. Enables a better randomness
savesignal = True
# ------ If savesignal is True, enter number of pseudo-period of superimposed
#        signals and repeat length
npseudoper = 30. # Number of pseudo period of superimposed signals.
len_repeat = 40000*14*50.
# ------ Roll error (True to compute it):
roll = True
# ------ Phase error (True to compute it):
phase = True
# ------ Baseline dilation error (True to compute it):
baseline_dilation = True
# ------ Timing error (True to compute it):
timing = True

## -- Geophysical error
## ----------------------
# ------ Wet tropo error (True to compute it):
wet_tropo = True
# ------ Beam print size (in km):
#        Gaussian footprint of sigma km
sigma = 8.
# ------ Number of beam used to correct wet_tropo signal (1, 2 or 'both'):
nbeam = 2
# ------ Beam position if there are 2 beams (in km from nadir):
beam_pos_l = -35.
beam_pos_r = 35.
# ------ Sea State Bias error (Not implemented yet):
ssb = False
# (ssb not implemented yet)
